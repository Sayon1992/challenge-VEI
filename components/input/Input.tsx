import { AutoComplete } from "antd";
import fetcher from "../../fetcher";
import React, { useState } from "react";
import { Coins } from "types/Coins";
import styles from "./Input.module.scss";
let timer: NodeJS.Timeout;

interface IInput {
  onSelect: (e: string, item: any) => void;
  reset: () => void;
}

const Input = ({ onSelect, reset }: IInput) => {
  const [options, setOptions] = useState<any[]>();
  const handleOptions = (data: Coins[]) => {
    const mappedOptions = data.map((i) => ({
      value: i.name,
      label: (
        <div className={styles.inputField}>
          <img className={styles.inputImage} src={i.iconUrl} />
          <h5>{i.name}</h5>
          <h5>{i.price}</h5>
        </div>
      ),
      uuid: i.uuid,
    }));
    setOptions(mappedOptions);
  };

  const handleSearch = async (value: string) => {
    if (value) {
      const { data } = await fetcher.get("/search-suggestions", {
        params: { query: value },
      });
      handleOptions(await data.data.coins);
      return;
    }
    setOptions([]);
    reset();
  };

  const onSearch = (v: string) => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      handleSearch(v);
    }, 300);
  };

  return (
    <AutoComplete
      className={styles.input}
      data-testid="autosuggest-trigger"
      options={options}
      style={{ width: 200 }}
      onSelect={onSelect}
      onSearch={onSearch}
      placeholder="input here"
    />
  );
};

export default Input;
