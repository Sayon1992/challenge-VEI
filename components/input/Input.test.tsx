import { fireEvent, render, waitFor } from "@testing-library/react";
import Input from "./Input";

describe("Input Component", () => {
  test("should match snapshot", () => {
    const { container } = render(<Input />);
    expect(container).toMatchSnapshot();
  });
  test("should show input and options of input when typed", async () => {
    const { getByTestId, getByText } = render(<Input />);
    const searchInput = getByTestId("autosuggest-trigger").querySelector(
      ".ant-select-selection-search-input"
    ) as HTMLInputElement;

    fireEvent.focus(searchInput!);

    fireEvent.change(searchInput!, { target: { value: "Bit" } });
    expect(searchInput?.value).toBe("Bit");

    fireEvent.click(getByText("Bitcoin"));

    await waitFor(() => {
      expect(document.querySelector(".ant-select-dropdown")).not.toBeNull();
    });
  });
});
