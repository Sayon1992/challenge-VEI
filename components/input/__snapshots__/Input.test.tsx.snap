// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component should match snapshot 1`] = `
<div>
  <div
    class="ant-select ant-select-auto-complete input ant-select-single ant-select-show-search"
    data-testid="autosuggest-trigger"
    style="width: 200px;"
  >
    <div
      class="ant-select-selector"
    >
      <span
        class="ant-select-selection-search"
      >
        <input
          aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
          aria-autocomplete="list"
          aria-controls="rc_select_TEST_OR_SSR_list"
          aria-haspopup="listbox"
          aria-owns="rc_select_TEST_OR_SSR_list"
          autocomplete="off"
          class="ant-select-selection-search-input"
          id="rc_select_TEST_OR_SSR"
          role="combobox"
          type="search"
          value=""
        />
      </span>
      <span
        class="ant-select-selection-placeholder"
      >
        input here
      </span>
    </div>
  </div>
</div>
`;
